File: observer/go/observer_test.go
package observer

import (
	"fmt"
	"testing"
	"time"
)

func TestObserver(t *testing.T) {
	var subject Subject = new(ConcreteSubject)
	subject.RegisterObserver(new(ConcreteObserverOne))
	subject.RegisterObserver(new(ConcreteObserverTwo))
	subject.NotifyObservers("hello world")

	fmt.Println("=====================================================")

	userController := NewUserController()
	promotion := RegPromotionObserver{}.HandleRegSuccess
	notification := RegNotificationObserver{}.HandleRegSuccess
	userController.SetRegObservers([]any{promotion, notification})
	userID := userController.Register("13800138000", "123!@#abc")
	fmt.Printf("1. observer_test.TestObserver: %d 注册成功\n", userID)
	userController.UnregObservers([]any{notification})
	userID = userController.Register("13800138001", "321!@#ABC")
	fmt.Printf("2. observer_test.TestObserver: %d 注册成功\n", userID)
	time.Sleep(300 * time.Millisecond)
}

File: observer/java/ObserverTest.java
import java.util.List;
import java.util.concurrent.TimeUnit;

public class ObserverTest {
    public static void main(String[] args) throws InterruptedException {
        Subject subject = new ConcreteSubject();
        subject.registerObserver(new ConcreteObserverOne());
        subject.registerObserver(new ConcreteObserverTwo());
        subject.notifyObservers("hello world");

        System.out.println("=====================================================");

        UserController userController = new UserController(new UserService());
        RegPromotionObserver promotion = new RegPromotionObserver();
        RegNotificationObserver notification = new RegNotificationObserver();
        userController.setRegObservers(List.of(promotion, notification));
        Long userId = userController.register("13800138000", "123!@#abc");
        System.out.printf("1. [%s] ObserverTest.main: %d 注册成功\n", Thread.currentThread(), userId);
        userController.unregObservers(List.of(notification));
        userId = userController.register("13800138001", "321!@#ABC");
        System.out.printf("2. [%s] ObserverTest.main: %d 注册成功\n", Thread.currentThread(), userId);
        TimeUnit.MILLISECONDS.sleep(300L);
    }
}
